bin/runapptests.js

```diff
@@ -71,97 +89,408 @@ var emu = require(BASE_DIR+"/core/lib/emulator.js");
-            break;
-          case "checkMemoryUsage" :
-            emu.tx(`\x10print(process.memory().usage)\n`);
-            var memUsage =  emu.getLastLine();
-            console.log("> CURRENT MEMORY USAGE", memUsage);
-            if (subtest.memUsage != memUsage ) {
-              console.log("> FAIL: EXPECTED MEMORY USAGE OF "+subtest.memUsage);
-              ok = false;
-            }
-            break;
-          default: ERROR("Unknown step type "+step.t);
-        }
-        emu.idle();
+        p = p.then(()=>{
+          return runStep(step, subtest, test, state).catch((e)=>{
+            console.log("> STEP FAILED:", e, step);
+            state.ok = false;
+          })
+        });
+      });
+
+      p = p.finally(()=>{
+        console.log("> RESULT -",  (state.ok ? "OK": "FAIL") , "- " + test.app + (subtest.description ? (" - " + subtest.description) : ""));
+        testState.push({
+          app: test.app,
+          number: subtestIdx,
+          result: state.ok ? "SUCCESS": "FAILURE",
+          description: subtest.description
+        });
       });
     });
-    emu.stopIdle();
+    p = p.then(()=>{
+      emu.stopIdle();
+    });
+    return p;
   });
 }
 
 
+let handleRx = ()=>{};
+let handleConsoleOutput = () => {};
+if (verbose){
+  handleConsoleOutput = (d) => {
+    console.log("<", d);
+  }
+}
+
+let testState = [];
+
 emu.init({
   EMULATOR : EMULATOR,
-  DEVICEID : DEVICEID
+  DEVICEID : DEVICEID,
+  rxCallback : (ch)=>{
+    handleRx(ch);
+  },
+  consoleOutputCallback: (d)=>{
+    handleConsoleOutput(d);
+  }
 }).then(function() {
   // Emulator is now loaded
   console.log("Loading tests");
-  var tests = [];
-  apploader.apps.forEach(app => {
-    var testFile = APP_DIR+"/"+app.id+"/test.json";
-    if (!require("fs").existsSync(testFile)) return;
-    var test = JSON.parse(require("fs").readFileSync(testFile).toString());
-    test.app = app.id;
-    tests.push(test);
+  let p = Promise.resolve();
+  let apps = apploader.apps;
+
+  apps.push(DEMOAPP);
+
+  if (process.argv.includes("--id")) {
+    let f = process.argv[process.argv.indexOf("--id") + 1];
+    apps = apps.filter(e=>e.id==f);
+    if (apps.length == 0){
+      console.log("No apps left after filtering for " + f);
+      process.exitCode(255);
+    }
+  }
+
+  apps.forEach(app => {
+    let test = DEMOTEST;
+    if (app.id != DEMOAPP.id){
+      let testFile = APP_DIR+"/"+app.id+"/test.json";
+      if (!require("fs").existsSync(testFile)) return;
+      test = JSON.parse(require("fs").readFileSync(testFile).toString());
+      test.app = app.id;
+    }
+    p = p.then(()=>{
+      return runTest(test, testState);
+    });
   });
-  // Running tests
-  runTest(TEST);
+  p.finally(()=>{
+    console.log("\n\n");
+    console.log("Overall results:");
+    console.table(testState);
+
+    process.exit(testState.reduce((a,c)=>{
+      return a + ((c.result == "SUCCESS") ? 0 : 1);
```

Comment:
This could potentially reach 256, in which case it's equivalent to `exit(0)`, perhaps we should:
```suggestion
      return a || ((c.result == "SUCCESS") ? 0 : 1);
```